flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/SingleThreadedIteratorSource.java

```diff
@@ -0,0 +1,196 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.flink.maintenance.operator;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.Collection;
+import java.util.Iterator;
+import org.apache.flink.api.connector.source.Source;
+import org.apache.flink.api.connector.source.SourceReader;
+import org.apache.flink.api.connector.source.SourceReaderContext;
+import org.apache.flink.api.connector.source.SplitEnumerator;
+import org.apache.flink.api.connector.source.SplitEnumeratorContext;
+import org.apache.flink.api.connector.source.lib.util.IteratorSourceEnumerator;
+import org.apache.flink.api.connector.source.lib.util.IteratorSourceReader;
+import org.apache.flink.api.connector.source.lib.util.IteratorSourceSplit;
+import org.apache.flink.api.java.typeutils.ResultTypeQueryable;
+import org.apache.flink.core.io.SimpleVersionedSerializer;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+
+/**
+ * Implementation of the Source V2 API which uses an iterator to read the elements, and uses a
+ * single thread to do so.
+ *
+ * @param <T> The return type of the source
+ */
+public abstract class SingleThreadedIteratorSource<T>
```

Comment:
I was thinking about adding this to the Flink source code in the future if this pans out for us.
I really hate the current Flink Source API which is very hard to read/understand for a simple source. I think we need to have an easier way to start creating the source.

Reply:
I am also wondering if `SingleThread` is needed. If users don't construct the source operator manually. the stitch code can set the operator parallelism as 1, like the `IcebergFilesCommitter`

Reply:
When using the new `Source` API, we need to have a single split, to archive the parallelism 1.

I don't expect that we need higher parallelism to archive our goals, and it results in simpler code if we rely on it,  to trigger Maintenance Tasks only once. Otherwise we need to find a way to drop TableChange events which are for snapshots which are committed before the snapshot used by an already triggered Maintenance Task.

Reply:
I think it is for the best if we preaggregate as many of the TableChanges as possible. Otherwise we might end up triggering unnecessary rewrites.
Let's say, we have S1 (2 new files), S2 (3 new files), S3 (2 new files), S4 (3 new files), and we have set that we need manifest file compaction after 2 commits.
If we end up reading the  snapshots parallel, then we might end up running the manifest rewrite task 4 times. The good solution would be to run it a single time.

I do not expect that the `MonitorSource` will need the parallelisation anyways. It would add complexity, and extra resource usage with no benefits.

Reply:
@stevenzwu I agree with @pvary that this is a valuable addition and I have worked closely together with him on these parts. 

I think the current implementation works very well for this use-case and has been tested extensively, complicating it with potential parallelism settings will only make it more error-prone. This can be changed in the future if we feel a strong need without any backward compatibility implications.